rust/gateway/src/eventloop.rs

```diff
@@ -0,0 +1,193 @@
+        Self {
+            tunnel,
+            control_rx,
+            portal,
+
+            // TODO: Pick sane values for timeouts and size.
+            connection_request_tasks: futures_bounded::FuturesMap::new(
+                Duration::from_secs(60),
+                100,
+            ),
+            add_ice_candidate_tasks: futures_bounded::FuturesSet::new(Duration::from_secs(60), 100),
+            print_stats_timer: tokio::time::interval(Duration::from_secs(10)),
+        }
+    }
+}
+
+impl Eventloop {
+    pub fn poll(&mut self, cx: &mut Context<'_>) -> Poll<Result<Infallible>> {
+        loop {
+            if let Poll::Ready(Some(ice_candidates)) = self.control_rx.poll_recv(cx) {
+                let _id = self.portal.send(
+                    "gateway",
+                    EgressMessages::BroadcastIceCandidates(ice_candidates),
+                );
+                continue;
+            }
+
+            match self.connection_request_tasks.poll_unpin(cx) {
+                Poll::Ready(((_, reference), Ok(Ok(gateway_rtc_session_description)))) => {
+                    let _id = self.portal.send(
+                        "gateway",
+                        EgressMessages::ConnectionReady(ConnectionReady {
+                            reference,
+                            gateway_rtc_session_description,
+                        }),
+                    );
+
+                    // TODO: If outbound request fails, cleanup connection.
+                    continue;
+                }
+                Poll::Ready(((client, _), Ok(Err(e)))) => {
+                    self.tunnel.cleanup_connection(client.into());
+                    let _ = self.tunnel.callbacks().on_error(&e);
+                    continue;
+                }
+                Poll::Ready(((client, reference), Err(e))) => {
+                    tracing::debug!(
+                        "Failed to establish connection {reference} from client {client:?}: {e}"
+                    );
+                    continue;
+                }
+                Poll::Pending => {}
+            }
+
+            match self.add_ice_candidate_tasks.poll_unpin(cx) {
+                Poll::Ready(Ok(Ok(()))) => {
+                    continue;
+                }
+                Poll::Ready(Ok(Err(e))) => {
+                    tracing::error!(err = ?e,"add_ice_candidate");
+                    let _ = self.tunnel.callbacks().on_error(&e);
+                    continue;
+                }
+                Poll::Ready(Err(e)) => {
+                    tracing::debug!("Failed to add ICE candidatee: {e}");
+                    continue;
+                }
+                Poll::Pending => {}
+            }
+
+            match self.portal.poll(cx)? {
+                Poll::Ready(phoenix_channel::Event::InboundMessage {
+                    msg: IngressMessages::Init(_),
+                    ..
+                }) => {
+                    todo!("received init after init");
```

Comment:
Can we just log an error here? We should use `ControlProtocolError` also we can add a debug assertion.

Reply:
Yeah these are WIP, thanks for pointing them out!

Reply:
Is a debug-assert by itself enough?

Reply:
I'd like to have some visibility if something like this happens in prod